src/cdomains/intDomain.ml

```diff
@@ -1853,27 +2405,156 @@ module IntDomTupleImpl = struct
+    , opt I4.refine_with_congruence d cong )
+
+  let refine_with_interval (a, b, c, d) intv =
+    let opt f =
+      curry @@ function Some x, y -> Some (f x y) | _ -> None
+    in
+    ( opt I1.refine_with_interval a intv
+    , opt I2.refine_with_interval b intv
+    , opt I3.refine_with_interval c intv
+    , opt I4.refine_with_interval d intv )
+
+  let refine_with_excl_list (a, b, c, d) excl =
+    let opt f =
+      curry @@ function Some x, y -> Some (f x y) | _ -> None
+    in
+    ( opt I1.refine_with_excl_list a excl
+    , opt I2.refine_with_excl_list b excl
+    , opt I3.refine_with_excl_list c excl
+    , opt I4.refine_with_excl_list d excl )
+
+  let refine_with_incl_list (a, b, c, d) incl =
+    let opt f =
+      curry @@ function Some x, y -> Some (f x y) | _ -> None
+    in
+    ( opt I1.refine_with_incl_list a incl
+    , opt I2.refine_with_incl_list b incl
+    , opt I3.refine_with_incl_list c incl
+    , opt I4.refine_with_incl_list d incl )
+
+
+    let mapp r (a, b, c, d) =
+    let map = BatOption.map in
+    ( map (r.fp (module I1)) a
+    , map (r.fp (module I2)) b
+    , map (r.fp (module I3)) c
+    , map (r.fp (module I4)) d)
+
+
+  let mapp2 r (a, b, c, d) =
+      BatOption.
+        ( map (r.fp2 (module I1)) a
+        , map (r.fp2 (module I2)) b
+        , map (r.fp2 (module I3)) c
+        , map (r.fp2 (module I4)) d )
+
+  let flat f x = match to_list_some x with [] -> None | xs -> Some (f xs)
+
+  let to_excl_list x = mapp2 { fp2 = fun (type a) (module I:S with type t = a and type int_t = int_t) -> I.to_excl_list } x |> flat List.concat
+  let to_incl_list x = mapp2 { fp2 = fun (type a) (module I:S with type t = a and type int_t = int_t) -> I.to_incl_list } x |> flat List.concat
+  let pretty () = (fun xs -> text "(" ++ (try List.reduce (fun a b -> a ++ text "," ++ b) xs with _ -> nil) ++ text ")") % to_list % mapp { fp = fun (type a) (module I:S with type t = a) -> (* assert sf==I.short; *) I.pretty () } (* NOTE: the version above does something else. also, we ignore the sf-argument here. *)
+
+  let refine_functions : (t -> t) list =
+    let maybe reffun domtup dom =
+      match dom with Some y -> reffun domtup y | _ -> domtup
+    in
+    [(fun (a, b, c, d) -> refine_with_incl_list (a, b, c, d) (to_incl_list (a, b, c, d)));
+     (fun (a, b, c, d) -> maybe refine_with_interval (a, b, c, d) b);
+     (fun (a, b, c, d) -> maybe refine_with_congruence (a, b, c, d) d)]
+
+  let refine ((a, b, c, d ) : t ) : t =
+    if not (GobConfig.get_bool "ana.int.refinement") then (a, b, c, d )
+    else
+      let dt = ref (a, b, c, d) in
+      dt := refine_with_excl_list !dt (to_excl_list (a, b, c, d));
+      let quit_loop = ref false in
+      while not !quit_loop do
+        let old_dt = !dt in
+        List.iter (fun f -> dt := f !dt) refine_functions;
+        quit_loop := old_dt = !dt;
+        if M.tracing then M.trace "cong-refine-loop" "old: %a, new: %a\n" pretty old_dt pretty !dt;
+      done;
```

Comment:
You can check in each step whether some of the domains became `bot`, and return the appropriate bottom value of `IntDomTuple` in this case.

Besides that, there should be an option to perform the refinement exactly once (so no further looping). The option ` "ana.int.refinement"` could thus be turned into a string, instead of a boolean, with possible values `"never"`, `"once"`, and `"fixpoint"`.